### PR TITLE
[foxy backport] Ensure rviz_common::MessageFilterDisplay processes messages in the main thread (#620)

### PR DESCRIPTION
**Pull request by spurnvoj**
Backport https://github.com/ros2/rviz/pull/620 to Foxy

### CODE REVIEW

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15203/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9923/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12883/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15384/)

**Comment by ahcorde** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15387/)

File: rviz_common/include/rviz_common/message_filter_display.hpp
Comment (jacobperron) [inline rviz_common/include/rviz_common/message_filter_display.hpp:183]: Removing a protected method is not API (or ABI) compatible.

File: rviz_common/include/rviz_common/ros_topic_display.hpp
Comment (jacobperron) [inline rviz_common/include/rviz_common/ros_topic_display.hpp]: Adding a virtual method is not ABI compatible

**Review by jacobperron — COMMENTED** [review]
I don't think we should accept this change since it is not ABI compatible (ie. it would require users of `MessageFilterDisplay` to re-compile). I'm not as concerned about the changes to `_RosTopicDisplay` since that is more of an implementation detail, but I think we should remain ABI compatible with `MessageFilterDisplay`.

**Comment by gbalke** [comment]
I've made some adjustments that are being PRed to @spurnvoj 's branch. Here are the results from my abi compatibility check:
```
:~/dev$ abi-dumper --compare ABI-1.dump ABI-2.dump
Added _Z17qRegisterMetaTypeISt10shared_ptrIKvEEiPKcPT_N9QtPrivate21MetaTypeDefinedHelperIS5_Xaasr12QMetaTypeId2IS5_E7DefinedntsrSA_9IsBuiltInEE11DefinedTypeE
Added _Z27qRegisterNormalizedMetaTypeISt10shared_ptrIKvEEiRK10QByteArrayPT_N9QtPrivate21MetaTypeDefinedHelperIS6_Xaasr12QMetaTypeId2IS6_E7DefinedntsrSB_9IsBuiltInEE11DefinedTypeE
Added _ZN11rviz_common16_RosTopicDisplay22typeErasedMessageTakenESt10shared_ptrIKvE
Added _ZN17QtMetaTypePrivate23QMetaTypeFunctionHelperISt10shared_ptrIKvELb1EE8DestructEPv
Added _ZN17QtMetaTypePrivate23QMetaTypeFunctionHelperISt10shared_ptrIKvELb1EE9ConstructEPvPS2_
```
image

Tested with the following commands in a docker container where `pkg_ws` contained `ros2/rviz:foxy` and `workspace` contained `gbalke/rviz:foxy_backport_620`:
```
$ colcon build --merge-install --packages-up-to rviz_common --cmake-args -DBUILD_TESTING=OFF --cmake-args -DCMAKE_CXX_FLAGS="-g -Og" -DCMAKE_C_FLAGS="-g -Og"

$ abi-dumper /home/ubuntu/pkg_ws/build/rviz_common/librviz_common.so -o ABI-1.dump -lver 1 -ld-library-path /home/ubuntu/dev/pkg_ws/build/rviz_ogre_vendor/ogre_install/lib/:/opt/ros/foxy/lib/

$ abi-dumper /home/ubuntu/workspace/build/rviz_common/librviz_common.so -o ABI-2.dump -lver 2 -ld-library-path /home/ubuntu/dev/pkg_ws/build/rviz_ogre_vendor/ogre_install/lib/:/opt/ros/foxy/lib/

$ abi-compliance-checker -l librviz_common -old ABI-1.dump -new ABI-2.dump
```

**Comment by spurnvoj** [comment]
@gbalke thanks for the changes, looks good. @jacobperron can you please review again the current version?

**Review by jacobperron — APPROVED** [review]
LGTM, thanks @gbalke for looking into an ABI compatible fix.

I'd like to get an extra +1 from one of the original reviewers to confirm that this is ABI-compatible.
@hidmic @clalancette @wjwwood  (thanks in advance!)

**Review by clalancette — APPROVED** [review]

**Comment by gbalke** [comment]
CI up to Rviz:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15339/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10050/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13019/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15533/)

**Comment by gbalke** [comment]
@jacobperron seems like the windows build is borked for unrelated reasons. I'll leave it to your best judgement when it comes to merging.

**Comment by ahcorde** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15539/)

**Comment by jacobperron** [comment]
Let's try Windows again: [Build Status](https://ci.ros2.org/job/ci_windows/15563/)

**Comment by ahcorde** [comment]
Windows warnings are not related.

**Pull request merged**
